lib/cimp/storage.go

```diff
@@ -28,46 +30,43 @@ func NewStorage(cfg Config) (*ConsulStorage, error) {
 	}, nil
 }
 
-func (cs *ConsulStorage) Save(kv KV) error {
+func (cs *ConsulStorage) Save(kv *KV) error {
 	var ops api.TxnOps
-	for k, v := range kv {
+
+	var i int
+	for key, path := range kv.Index {
+		i++
+		leaf, err := kv.get(path)
+		if err != nil {
+			return fmt.Errorf("get key %q value from tree: %w", key, err)
+		}
+
 		op := &api.TxnOp{
 			KV: &api.KVTxnOp{
 				Verb:  api.KVSet,
-				Key:   k,
-				Value: []byte(fmt.Sprintf("%v", v)),
+				Key:   kv.GlobalPrefix + key,
+				Value: []byte(fmt.Sprintf("%v", leaf.Value)),
 			},
 		}
 		ops = append(ops, op)
+
+		if len(ops) == consulTransactionLimit || i == len(kv.Index) {
+			if ok, _, _, err := cs.client.Txn().Txn(ops, nil); !ok {
+				return fmt.Errorf("execute consul SET-transaction: %w", err)
+			}
+			ops = nil
+		}
 	}
 
-	ok, _, _, err := cs.client.Txn().Txn(ops, nil)
-	if !ok {
+	if ok, _, _, err := cs.client.Txn().Txn(ops, nil); !ok {
 		return fmt.Errorf("execute consul SET-transaction: %w", err)
```

Comment:
does err  != nil always for ok==true ?

Reply:
// If there is a problem making the transaction request then an error will be
// returned. Otherwise, the ok value will be true if the transaction succeeded
// or false if it was rolled back.

Part of documentation

Reply:
for ok == false, not true